apps/ec2-unattached-ebs-volumes/app.py

```
@@ -22,5 +22,8 @@ def lambda_handler(event, context):
 
     response = client.execute(query)
     response_json = json.loads(response)
-    volumes = response_json["data"]["ec2_volume"]["data"]
-    return [v for v in volumes if not v["attachments"]]
+    if (response_json["data"]):
+        volumes = response_json["data"]["ec2_volume"]["data"]
+        return [v for v in volumes if not v["attachments"]]
+    else:
+        return response_json["errors"]
```

Is the idea that we defer error handling to the clients. My issue here is with swallowing errors silently especially if it returns the same base data struct as the response. One idea is to do what we generally do for APIs and have a result JSON like: 
Success case:
```
{
  "status": "OK",
  "data": [ { JSON } ],
  "errors": [JSON]
}```
Failure
```
{
   "status": "ERROR_CODE",
  "data": []
  "errors": [JSON]
}

or without the `errors` and `data` fields for the `Success` and `Failure` cases

---

There's a tradeoff here. Adding wrapping objects adds inconvenience for the common happy case, in particular when piping apps.

+ @chalenge's suggestion fits well with the GraphQL way of thinking, in particular its [error handling](https://spec.graphql.org/June2018/#sec-Response-Format). However it's not Unix-y: piping from one app to the other will require intermediate steps to unwrap the data: each `|` will become `| jq '.data' |` (this would get bigger with error handling).
+ A more Unix-oriented solution could instead expose failures to the client by 1) returning with non-0 exit code and 2) writing data to `stdout` and errors to `stderr`. This will ensure pipes work as expected, including termination.

My recommendation would be that the server API return responses in the GraphQL format (option 1) and the CLI do so in the Unix format (option 2).

---

So, just to confirm, apps (and api) should return {data:{}, errors:{}}. Cli will inspect and transform.

---

It would be nice if we could keep returning unwrapped data and throwing exceptions as usual from the apps. We also can't guarantee that apps won't throw exceptions and it would be nice to handle that case consistently.

Reading [this page](https://docs.aws.amazon.com/lambda/latest/dg/python-exceptions.html) I _think_ it's possible to have the behavior we want in the API and CLI while keeping the CLI as is, by inspecting the data returned by AWS.

---

I actually suggest this a part 1 of a more comprehensive error strategy beyond the graphql part. I also do not see the apps always being CLI based.